README.md
# partCNV
An R package written by **Dr. Ziyi Li** to improve the detection of locally aneuploid cells by incorporating cytogenetics information.

# Introduction

Single-cell RNA sequencing is becoming an increasingly common tool to investigate the cellular population and patients' outcome in cancer research. However, due to the sparse data and the complex tumor microenvironment, it is challenging to identify neoplastic cells that play important roles in tumor growth and disease progression. This challenge is exaggerated in the research of blood cancer patients, from whom the neoplastic cells can be highly similar to the normal cells.

In this package we present partCNV/partCNVH, a statistical framework for rapid and accurate detection of aneuploid cells with local copy number deletion or amplification. Our method uses an EM algorithm with mixtures of Poisson distributions while incorporating **cytogenetics information** to guide the classification (partCNV). When applicable, we further improve the accuracy by integrating a Hidden Markov Model for feature selection (partCNVH).


# Understand your cytogenetics data

First of all, our method can only be used to infer cell status if the patient has a part of the cells with cytogenetics alterations. For example, if a patient has cytogenetics data as ``46,XY,del(20)(q11.1q13.1)[5]/46,XY[15]``, it means that out of 20 metaphases, 5 (i.e., 25%) of the cells has deletion on chromosome 20 in region q11.1 to q13.1. If the patient has a normal cytogenetics, e.g., 46,XY[20], or all altered cells, e.g., 46,XY,del(20)(q11.1q13.1)[25], there won't be any need to apply the proposed method. 

Second, when you have a complicated cytogenetics feature, use them one by one to identify the desired cell group. For example, ``47,XY,+8[5]/46,idem,del(8)(q21.2q24.3)/46,XY[7]``, we start with the chromosome amplification on chromosome 8 excluding the region (q21.2q24.3) to identify the cells with this alteration using the proposed method. After the cells with chromosome 8 amplification are identified, we apply the proposed method to identify cells with del(8)(q21.2q24.3). 

Lastly, our tool is primarily designed for analyzing the scRNA-seq data from patients with hematologic malignancies, such as MDS and AML. For solid tumors such as lung cancer and breast cancer, it is better to use CNV based method such as [copyKAT](https://github.com/navinlabcode/copykat) and inferCNV `r Biocpkg("inferCNV")`.


# Load example data

The analysis can start from whole-genome scRNA-seq data or a subset of scRNA-seq data based on the location of interest. In this package, we prepared a whole-genome scRNA-seq data:


<!-- Assign captions to figures in the code chunk option `fig.cap` to automatically number them, and to be able to reference them, see Figure \@ref(fig:plot). The figure label is generated from the code chunk label by prefixing it with `fig:`. -->

```{r loadData, echo=TRUE}
library(partCNV)
data(SimData)
dim(SimData)
SimData[1:5,1:5]
```

This simple example dataset contains 24519 genes and 400 cells. If you start with Seurat object (after quality control), run:

```{r runseurat, echo=TRUE, eval=FALSE}
library(Seurat)
Seurat_obj <- NormalizeData(Your_SeuratObj, normalization.method = "LogNormalize", scale.factor = 10000)
Counts = Seurat_obj@assays$RNA@counts
```

# Run partCNV

Since this is simulation data, prior knowledge (e.g., cytogenetics data in real studies) shows that this example data has 40% of cells with deletion on chromosome 20 q11.1 to q13.1. Let's get started with locating this region.

```{r s1, echo=TRUE, eval=TRUE}
res <- GetCytoLocation(cyto_feature = "chr20(q11.1-q13.1)")
```

Then let's subset the data and normalize the data:
```{r s2, echo=TRUE, eval=TRUE}
GEout <- GetExprCountCyto(cytoloc_output = res, Counts = SimData, normalization = TRUE, qt_cutoff = 0.99)
```
For this function, the qt_cutoff is to filter out the cells with very low expressions. 0.99 here means that we filter out cells that only express in 1% (1-0.99) cells. Make this `qt_cutoff` larger if your total gene number within the region is small. 

Now we apply partCNV:
```{r s3, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, results='hide'}
pcout <- partCNV(int_counts = GEout$ProcessedCount,
                 cyto_type = "del",
                 cyto_p = 0.40)
```

Understand the results:
```{r s4, echo=TRUE, eval=TRUE}
table(pcout)
sum(pcout==1)/length(pcout)
p1 <- sum(pcout==1)/length(pcout)
```
39.5% of cells are labeled as locally aneuploid (1) and others are diploid (0).

Let's visualize it:
```{r s5, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, results='hide'}
library(Seurat)
sim_seurat <- CreateSeuratObject(counts = SimData)
sim_seurat <- NormalizeData(sim_seurat, normalization.method = "LogNormalize", scale.factor = 10000)
sim_seurat <- FindVariableFeatures(sim_seurat, selection.method = "vst", nfeatures = 2000)
all.genes <- rownames(sim_seurat)
sim_seurat <- ScaleData(sim_seurat, features = all.genes)
sim_seurat <- RunPCA(sim_seurat, features = VariableFeatures(object = sim_seurat))
sim_seurat <- RunUMAP(sim_seurat, dims = 1:10)
sim_seurat <- AddMetaData(
    object = sim_seurat,
    metadata = pcout,
    col.name = "partCNV_label"
)
sim_seurat$partCNV_label <- factor(sim_seurat$partCNV_label, levels = c(1,0), labels = c(
    "aneuploid", "diploid"
))
```

```{r s6, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE}
library(ggplot2)
DimPlot(sim_seurat, reduction = "umap", group = "partCNV_label") + ggtitle(paste0("partCNV (", signif(p1,2)*100, "%)"))+ theme(legend.position = "none")
```

# Run partCNVH

Compared with partCNV, partCNVH added an additional step of feature selection. This is especially helpful if your cytogenetics provide a very broad region and part of it does not have chromosomal alterations. The first two steps of using partCNVH are exactly the same as using partCNV.


```{r s11, echo=TRUE, eval=TRUE}
res <- GetCytoLocation(cyto_feature = "chr20(q11.1-q13.1)")
GEout <- GetExprCountCyto(cytoloc_output = res, Counts = SimData, normalization = TRUE, qt_cutoff = 0.99)
```

For this function, the qt_cutoff is to filter out the cells with very low expressions. 0.99 here means that we filter out cells that only express in 1% (1-0.99) cells. Make this `qt_cutoff` larger if your total gene number within the region is small. 

Now we apply partCNVH:
```{r s13, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, results='hide'}
pcHout <- partCNVH(int_counts = GEout$ProcessedCount,
                 cyto_type = "del",
                 cyto_p = 0.40)
```

Understand the results (in pcHout, EMlabel is the partCNV label and EMHMMlabel is the partCNVH label).
```{r s14, echo=TRUE, eval=TRUE}
table(pcHout$EMHMMlabel)
sum(pcHout$EMHMMlabel==1)/length(pcHout$EMHMMlabel)
p2 <- sum(pcHout$EMHMMlabel==1)/length(pcHout$EMHMMlabel)
```
41.25% of cells are labeled as locally aneuploid (1) and others are diploid (0).

Let's visualize it:
```{r s15, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, results='hide'}
# I commented these steps because they are exactly the same as partCNV run. 
# library(Seurat)
# sim_seurat <- CreateSeuratObject(counts = SimData)
# sim_seurat <- NormalizeData(sim_seurat, normalization.method = "LogNormalize", scale.factor = 10000)
# sim_seurat <- FindVariableFeatures(sim_seurat, selection.method = "vst", nfeatures = 2000)
# all.genes <- rownames(sim_seurat)
# sim_seurat <- ScaleData(sim_seurat, features = all.genes)
# sim_seurat <- RunPCA(sim_seurat, features = VariableFeatures(object = sim_seurat))
# sim_seurat <- RunUMAP(sim_seurat, dims = 1:10)
sim_seurat <- AddMetaData(
    object = sim_seurat,
    metadata = pcHout$EMHMMlabel,
    col.name = "partCNVH_label"
)
sim_seurat$partCNVH_label <- factor(sim_seurat$partCNVH_label, levels = c(1,0), labels = c(
    "aneuploid", "diploid"
))
```

```{r s16, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE}
library(ggplot2)
DimPlot(sim_seurat, reduction = "umap", group = "partCNVH_label") + ggtitle(paste0("partCNVH (", signif(p2,2)*100, "%)"))+ theme(legend.position = "none")
```


NEWS.md
## PartCNV 0.99.0 (2022-1-30)

### Changes
- Added unit tests.


R/partCNVH.R
#' Infer cells that are locally aneuploid using partCNVH
#'
#' This function uses EM algorithm to cluster the cells with a Poisson Mixture model in the first step.
#' With the results, it applies hidden markov model to improve feature selection.
#' After that, another round of EM algorithm is applied to obtain the final cell status.
#' Cells will be grouped into two groups, locally aneuploid (status = 1) and
#' diploid (status = 0).
#'
#' @param int_counts Normalized gene expression counts for the genes in the interested
#' region, e.g., the ProcessedCount variable from the output of GetExprCountCyto().
#' @param cyto_type The type of the cytogenetics alteration. It can only be "del" or "amp"
#' @param cyto_p The percentage of cells with the cytogenetics alteration, e.g., 0.2.
#' @param tau The variance of the prior information. Default is 0.1. If you have less confidence,
#' specify a larger tau, e.g., 10.
#' @param maxniter The maximum number of iterations of the EM algorithm.
#' @param navg Number of genes used for rolling average.
#'
#' @return A vector with the cell status inferred by the method, 1 is aneuploid and 0 is diploid.
#' @export
#' @examples
#' cytoloc <- GetCytoLocation(cyto_feature = "chr20(q11.1-q13.1)")
#' data(SimData)
#' exprout <- GetExprCountCyto(cytoloc_output = cytoloc, Counts = as.matrix(SimData), normalization = TRUE, qt_cutoff = 0.99)
#' status <- partCNVH(int_counts = exprout$ProcessedCount, cyto_type = "del", cyto_p = 0.2, navg = 50)
#'
partCNVH <- function(int_counts,
                     cyto_type,
                     cyto_p,
                     tau = 0.1,
                     maxniter = 1000,
                     navg = 50) {

    if (nrow(int_counts) <= 20) {
        warning("The number of genes is only ", nrow(int_counts), ". It maybe better to use partCNV only.")
    }

    EMlabel <- partCNV(int_counts = int_counts,
                     cyto_type = cyto_type,
                     cyto_p = cyto_p,
                     tau = tau,
                     maxniter = maxniter)

    if(cyto_type == "del") {
        meanratio <- rowMeans(int_counts[, EMlabel == 0])/rowMeans(int_counts[, EMlabel == 1])
        meanratio2 <- data.table::frollmean(meanratio, n = navg, na.rm = TRUE, align = "center")
        mysumdata <- data.frame(rowmean = meanratio2)
        initStatus <- rep(1, length(mysumdata$rowmean))
        initStatus[mysumdata$rowmean > stats::median(mysumdata$rowmean)] <- 2
        mod <- depmixS4::depmix(rowmean ~ 1, data = mysumdata, nstates = 2, initdata = initStatus, trstart = c(0.9,0.1,0.1,0.9)) # use gaussian() for normally distributed data
        fit.mod <- depmixS4::fit(mod)
        est.states <- depmixS4::posterior(fit.mod)

        if(mean(mysumdata$rowmean[est.states$state == 2], na.rm = TRUE) < mean(mysumdata$rowmean[est.states$state == 1], na.rm = TRUE)) {
            myidealstate <- 1
        } else {
            myidealstate <- 2
        }
    } else if(cyto_type == "amp") {
        meanratio <- rowMeans(int_counts[, EMlabel == 1])/rowMeans(int_counts[, EMlabel == 0])
        meanratio2 <- data.table::frollmean(meanratio, n = navg, na.rm = TRUE, align = "center")
        mysumdata <- data.frame(rowmean = meanratio2)
        initStatus <- rep(1, length(mysumdata$rowmean))
        initStatus[mysumdata$rowmean > stats::median(mysumdata$rowmean)] <- 2
        mod <- depmixS4::depmix(rowmean ~ 1, data = mysumdata, nstates = 2, initdata = initStatus, trstart = c(0.9,0.1,0.1,0.9)) # use gaussian() for normally distributed data
        fit.mod <- depmixS4::fit(mod)
        est.states <- depmixS4::posterior(fit.mod)

        if(mean(mysumdata$rowmean[est.states$state == 2], na.rm = TRUE) < mean(mysumdata$rowmean[est.states$state == 1], na.rm = TRUE)) {
            myidealstate <- 1
        } else {
            myidealstate <- 2
        }
    }

    EMHMMlabel <- partCNV(int_counts[est.states$state == myidealstate, ],
                        cyto_type = cyto_type,
                        cyto_p = cyto_p,
                        tau = tau,
                        maxniter = maxniter)

    return(list(EMlabel = EMlabel,
                EMHMMlabel = EMHMMlabel))
}




R/GetCytoLocation.R
#' Get exact location of the interested cytogenetics feature
#'
#' This function helps you identify the location of the cytogenetics feature.
#' For example, if the region of interest is chr20(q11.1-q13.1), this function
#' greps the start and end location of this
#' region. Additionally, you can just put in "chr20", and it provides you all
#' the available cytogenetics locations on chromosome 20.
#' It also report the number of genes within the region. If the number of genes
#' is too few, we recommend to include neighboring regions to provide more
#' stable results.
#'
#' @param cyto_feature the cytogenetics location you are interested. It can be
#' of two format: chr20(q11.1-q13.1) or chr20. For the first format, the start
#' and end regions need to be separated by "-". If you are interested in one
#' region for example, chr20(q11.1), put it as chr20(q11.1-11.1). For the second
#' format, all the available regions will be printed for selection.
#' @param chr chromosome location of the interested region. This is only used
#' when cyto_feature is null.
#' @param start starting location of the interested region. This is only used
#' when cyto_feature is null.
#' @param end ending location of the interested region. This is only used
#' when cyto_feature is null.
#' 
#' @import AnnotationHub
#'
#' @return If the first format of cyto_feature is provided, the starting and
#' ending location as well as the number of genes overlapped with be provided.
#' If the second format of cyto_feature is provided, all the cytogenetics locations
#' will be displayed for review. If the region location (chr, start, end) is
#' provided, the number of genes overlapped will be the output.
#' @export
#' @examples
#' ### example 1
#' GetCytoLocation(cyto_feature = "chr20(q11.1-q13.1)")
#' ### example 2
#' GetCytoLocation(cyto_feature = "chr20")
#' ### example 3
#' GetCytoLocation(chr = "chr20", start = 25600000, end = 49800000)
#'
GetCytoLocation <- function(cyto_feature = NULL,
                            chr = NULL,
                            start = NULL,
                            end = NULL) {

    gene <- data.frame(AnnotationHub()[["AH53178"]])
    cytoBand_Hg38 <- gene[, c("seqnames", "start", "end", "name", "gieStain")]
    colnames(cytoBand_Hg38) <- c("chr", "start", "end", "cytoband", "V5")  
    
    Hg38_gtf <- NULL
    data("Hg38_gtf", envir=environment())

    GeneNum_index <- 0
    Downstream_index <- 0

    if(is.null(cyto_feature) & is.null(chr)) {
        stop("If cyto_feature is null, please provide the region location through chr, start, and end!")
    } else if (!is.null(cyto_feature)) {

        if(length(grep("\\(", cyto_feature)) == 0) {

            if(length(grep(cyto_feature, cytoBand_Hg38$chr)) == 0) {
                stop("Please provide a valid cyto_feature input!")
            } else {
                message("Print out all cytogenetics features on ", cyto_feature, ":")
                tmpout <- cytoBand_Hg38[grep(cyto_feature, cytoBand_Hg38$chr), ]
                rownames(tmpout) <- NULL
                message(tmpout)
                return(list(cytogeneticsInfo = tmpout,
                            Downstream_index = Downstream_index))
            }

        } else if (length(grep("\\(", cyto_feature)) > 0) {

            tmpout2 <- unlist(strsplit(cyto_feature, split = "\\("))
            chr_location <- tmpout2[1]
            if(length(grep(chr_location, cytoBand_Hg38$chr)) == 0) {
                stop("Please provide a valid chromosome information in cyto_feature input!")
            } else {
                filtered_cyto <- cytoBand_Hg38[grep(chr_location, cytoBand_Hg38$chr), ]
            }

            tmpout3 <- gsub(")", "", tmpout2[2])
            tmpout4 <- unlist(strsplit(tmpout3, split = "-"))

            idx1 <- grep(tmpout4[1], filtered_cyto$cytoband)
            idx2 <- grep(tmpout4[2], filtered_cyto$cytoband)

            if(length(idx1) == 0 | length(idx2) == 0) {
                stop("The cytoband information couldn't be recognized. Please try using the chr info only.")
            } else {
                cyto_start <- min(filtered_cyto$start[idx1])
                cyto_end <- max(filtered_cyto$end[idx2])
            }
            GeneNum_index <- 1

        }
    } else if(is.null(cyto_feature) & !is.null(chr) & !is.null(start) & !is.null(end)) {

        chr_location <- chr
        cyto_start <- start
        cyto_end <- end
        GeneNum_index <- 1

    } else {
        stop("The chr/start/end information couldn't be recognized.")
    }

    if(GeneNum_index == 1) {

        gene.gr <- GenomicRanges::GRanges(seqnames = Hg38_gtf$seqnames,
                                          ranges = IRanges::IRanges(start = as.numeric(Hg38_gtf$start),
                                                                    end = as.numeric(Hg38_gtf$end)))
        cyto.gr <- GenomicRanges::GRanges(seqnames = chr_location,
                                          ranges = IRanges::IRanges(start = cyto_start,
                                                                    end = cyto_end))
        FOout <- as.data.frame(GenomicRanges::findOverlaps(cyto.gr, gene.gr))
        overGene <- unique(Hg38_gtf$gene_id[unique(FOout$subjectHits)])
        overGeneName <- unique(Hg38_gtf$gene_name[unique(FOout$subjectHits)])
        message("Interested region: ", chr_location, ":", cyto_start, "-", cyto_end, ".")
        message("A total of ", length(overGeneName), " genes are located in this region.")
        Downstream_index <- 1

        return(list(chr_location = chr_location,
                    cyto_start = cyto_start,
                    cyto_end = cyto_end,
                    overGeneID = overGene,
                    overGeneName = overGeneName,
                    Downstream_index = Downstream_index))
    }
}




R/GetExprCountCyto.R
#' Get normalized gene expression counts for selected genes
#'
#' This function helps normalize the gene expression count matrix if needed
#' and select the genes that are located in the interested region. This procedure
#' happens after applying GetCytoLocation().
#'
#' @param cytoloc_output The output from the function GetCytoLocation().
#' The function needs to be run with a complete cytogenetics feature input, e.g.,
#' chr20(q11.1-11.1), or providing the chr/start/end loction of the interested
#' region.
#' @param Counts The single cell expression matrix for the whole genome of the
#' sample. Rows are genes and columns are cell IDs.
#' @param normalization Specify whether the data need to be normalized. Default is TRUE.
#' @param qt_cutoff A quantile cut-off to remove genes that are almost all zeros.
#' If the cut-off is 0.99, then all the genes expressed in less than 0.01 percent
#' of cells will be eliminated for further analysis.
#'
#' @return A list with normalized and ordered gene expression for the interested
#' cytogenetics region.
#' @import Matrix
#' @export
#' @examples
#' res <- GetCytoLocation(cyto_feature = "chr20(q11.1-q13.1)")
#' data(SimData)
#' GetExprCountCyto(cytoloc_output = res, Counts = as.matrix(SimData), normalization = TRUE, qt_cutoff = 0.99)
#'
GetExprCountCyto <- function(cytoloc_output,
                             Counts = NULL,
                             normalization = TRUE,
                             qt_cutoff = 0.99) {

    Hg38_gtf <- NULL

    data("Hg38_gtf", envir=environment())

    if(cytoloc_output$Downstream_index == 0) {
        stop("Please re-run GetCytoLocation with a complete cyto input, e.g., chr20(q11.1-q13.1), or provide chr/start/end information.")
    } else if(cytoloc_output$Downstream_index == 1) {

        if(!normalization) {
            normalizedCounts <- Counts
        } else {
            normFactor <- colMeans(Counts)/stats::median(colMeans(Counts))
            normalizedCounts <- sweep(Counts, 2, normFactor, "/")
            rownames(normalizedCounts) <- rownames(Counts)
        }

        fCounts <- normalizedCounts[stats::na.omit(match(cytoloc_output$overGeneName, rownames(normalizedCounts))), ]
        tmpGL <- Hg38_gtf[match(rownames(fCounts), Hg38_gtf$gene_name), ]
        GeneLocation <- tmpGL$start
        fCounts_ord <- fCounts[order(GeneLocation),]
        GeneLocation_ord <- GeneLocation[order(GeneLocation)]

        fout <- Fselect(fCounts_ord, cutoff = qt_cutoff, status = GeneLocation_ord)

        return(list(ProcessedCount = fout$count,
                    status = fout$status))
    }

}


R/partCNV.R
#' Infer cells that are locally aneuploid using partCNV
#'
#' This function uses EM algorithm to cluster the cells with a Poisson Mixture model.
#' Cells will be grouped into two groups, locally aneuploid (status = 1) and
#' diploid (status = 0).
#'
#' @param int_counts Normalized gene expression counts for the genes in the interested
#' region, e.g., the ProcessedCount variable from the output of GetExprCountCyto().
#' @param cyto_type The type of the cytogenetics alteration. It can only be "del" or "amp"
#' @param cyto_p The percentage of cells with the cytogenetics alteration, e.g., 0.2.
#' @param tau The variance of the prior information. Default is 0.1. If you have less confidence,
#' specify a larger tau, e.g., 10.
#' @param maxniter The maximum number of iterations of the EM algorithm.
#'
#' @return A vector with the cell status inferred by the method, 1 is aneuploid and 0 is diploid.
#' @export
#' @examples
#' ### example 1
#' cytoloc <- GetCytoLocation(cyto_feature = "chr20(q11.1-q13.1)")
#' data(SimData)
#' exprout <- GetExprCountCyto(cytoloc_output = cytoloc, Counts = as.matrix(SimData), normalization = TRUE, qt_cutoff = 0.99)
#' status <- partCNV(int_counts = exprout$ProcessedCount, cyto_type = "del", cyto_p = 0.2)
#'
partCNV <- function(int_counts,
                  cyto_type,
                  cyto_p,
                  tau = 0.1,
                  maxniter = 1000) {

    ngene <- nrow(int_counts)
    ncell <- ncol(int_counts)

    tmp <- int_counts
    int_counts <- round(as.matrix(int_counts))

    ### initialization with PCA+louvian
    input <- int_counts
    if(is.null(rownames(int_counts))) {
        rownames(input) <- paste0("gene", seq_len(nrow(input)))
    }
    if(is.null(colnames(int_counts))) {
        colnames(input) <- paste0("cell", seq_len(ncol(input)))
    }
    sim <- Seurat::CreateSeuratObject(counts = input, project = "sim", min.cells = 0, min.features = 0)
    sim <- Seurat::NormalizeData(sim)
    sim <- Seurat::ScaleData(sim, features = rownames(input))
    sim <- Seurat::RunPCA(sim, features = rownames(input))
    sim <- Seurat::FindNeighbors(sim, dims = seq_len(10))
    reso_vec <- seq(0.001, 0.5, by = 0.01)
    K <- 1
    j <- 1
    while(K == 1) {
        sim <- Seurat::FindClusters(sim, resolution = reso_vec[j], algorithm = 1)
        j <- j + 1
        K <- length(unique(Seurat::Idents(sim)))
    }
    cellZ <- Seurat::Idents(sim)

    mres <- tapply(colMeans(int_counts), cellZ, mean)
    qi <- sum(cellZ == names(which.min(mres)))/length(cellZ)
    choice1 <- cutfunc(abs(stats::rnorm(length(cellZ), mean = qi, sd = 0.1)))
    choice0 <- cutfunc(abs(stats::rnorm(length(cellZ), mean = 1-qi, sd = 0.1)))
    pi <- ifelse(cellZ == names(which.min(mres)), choice1, choice0)
    mu_g1 <- rowSums(int_counts * getweightmat(pi, ngene, ncell))/sum(pi)
    mu_g0 <- rowSums(int_counts * getweightmat(1-pi, ngene, ncell))/sum(1-pi)
    mu_g1[mu_g1<0] <- 0.01
    mu_g0[mu_g0<0] <- 0.01

    ### q0 is always the probability of group with lower gene expressions
    if (tolower(cyto_type) == "del") {
        q0 <- cyto_p
    } else if (tolower(cyto_type) == "amp") {
        q0 <- 1 - cyto_p
    } else {
        stop("cyto_type must be del or amp!")
    }

    pi_old <- rep(0, length(cellZ))
    diff_p <- 10
    niter <- 1
    Ci <- rep(0, length(cellZ))
    diff_prp <- 10
    ### update the parameters
    while(diff_p > 10^-5 & diff_prp > 10^-5 & niter < maxniter) {
        pi_old <- pi
        Cold <- Ci

        # ## mu_g1 group should be the group with smaller mean
        if(mean(mu_g1) > mean(mu_g0)) {
            tmp <- mu_g0
            mu_g0 <- mu_g1
            mu_g1 <- tmp
            qi <- 1-qi
        }

        # ## update qi
        Ci <- ifelse(pi>=0.5, 1, 0)
        Ctotal <- sum(Ci)
        qi <- Ctotal/ncell
        fnx <- function(x) {
            Ctotal/x - (ncell - Ctotal)/(1-x) - ncell*(x - q0)/tau^2
        }
        rrout <- stats::uniroot(fnx, interval = c(0.001,0.99))
        qi <- rrout$root

        ## update qi
        logP1 <- stats::dpois(as.matrix(int_counts), lambda = mu_g1, log=TRUE)
        logP1sum <- colSums(logP1)
        logP0 <- stats::dpois(as.matrix(int_counts), lambda = mu_g0, log=TRUE)
        logP0sum <- colSums(logP0)
        pi <- qi/(qi + exp(logP0sum - logP1sum)*(1-qi))

        ## update mu
        mu_g1 <- rowSums(int_counts * getweightmat(pi, ngene, ncell))/sum(pi)
        mu_g0 <- rowSums(int_counts * getweightmat(1-pi, ngene, ncell))/sum(1-pi)

        Ci <- ifelse(pi>=0.5, 1, 0)
        Ctotal <- sum(Ci)

        if(Ctotal == 0 | Ctotal == ncol(int_counts)) {
            cutoff <- stats::quantile(pi, qi)
            Ci <- ifelse(pi >= cutoff, 1, 0)
            Ctotal <- sum(Ci)
        }

        niter <- niter + 1
        diff_p <- abs(sum(pi - pi_old))
        diff_prp <- sum(abs(Ci - Cold))/ncell
        message("qi = ", qi)
        message("diff_p = ", diff_p)
        message(Ctotal/ncell)
    }

    if (tolower(cyto_type) == "amp") {
        Ci <- 1 - Ci
        message(sum(Ci)/ncell)
    }

    return(Ci)
}
